### DISCUSSION TITLE
JAX for Production?

### DISCUSSION THREAD

**Discussion by pmh722** [jax-ml/jax#7120] [Q&A]
What is the current thinking on using JAX in production? Is it still bleeding edge research software? What’s the expected pace of breaking changes?

I experimented with JAX a year ago and loved it. Performance was great. API was logical and concise. But it seemed like I constantly had to pin JAX and XLA versions to keep code from breaking. 

I’m about to start a new project and am deciding between JAX and PyTorch. I’d prefer JAX but am worried about the pace of breaking change. 

To be clear I fully understand and support the JAX community continuing to iterate on the API until it is perfect. I have tremendous respect for the team and what they’ve built. I may just be a year or two early for my needs. 

Thanks very much for any insights you can offer.

**Comment by georgedahl** [comment]
Do breaking changes matter if the the version number is bumped appropriately? Can you clarify a bit more what you need from a library in this regard? Are you looking for some sort of policy statement? Or something else?

**Reply by pmh722** [reply]
Thanks for replying. 

My goal is to write software that will be embedded in a product and therefore needs to keep running for years with the minimal amount of additional developer effort to keep it running. 

Downsides of just pinning on one JAX / Jaxlib /FLAX version forever are 1) organizations gets squeamish for security reasons if we have a policy of never updating production dependencies, and 2) stagnating capabilities while the ecosystem continues to innovate. 

A policy of no-breaking-changes, either in the main branch or some LTS branch, would be optimal, but I don’t even need that. If breaking changes were small and infrequent we could handle it. 

Again, none of this is meant as a critique or even a request. If the JAX devs think the API is still far from a global optimum, they should keep the pace of change high until everything is optimal. I just wanted to check and see if we were close to converging :)

**Reply by georgedahl** [reply]
Does PyTorch have some sort of no-breaking-changes policy? I'm not familiar with it.

TensorFlow's policy on versioning seems to be at https://www.tensorflow.org/guide/versions but from my own experience it is still constantly changing.

**Reply by pmh722** [reply]
https://pytorch.org/docs/stable/index.html
"We also expect to maintain backwards compatibility (although breaking changes can happen and notice will be given one release ahead of time)."

They also have an open source Long Term Support version, which I believe pushes bug fixes but otherwise tries as hard as possible to keep everything stable. https://pytorch.org/enterprise-support-program

I care less about the language of the policy then I do about the empirical frequency and magnitude of breaking changes. In general I do find PyTorch to be pretty stable ever since they went 1.0. I realize this is partially subjective and depends on which parts of the API you depend on. But my impression (a year ago) is that JAX had more frequent and bigger breaking changes than PyTorch. (Again, not saying that's the wrong call for the JAX community at this stage in its development.)
